File: src/main/java/net/frozenorb/potpvp/util/serialization/LocationAdapter.java
package net.frozenorb.potpvp.util.serialization;

import com.google.gson.*;
import net.frozenorb.potpvp.PotPvPRP;
import org.bukkit.Location;
import org.bukkit.World;

import java.lang.reflect.Type;

public class LocationAdapter implements JsonDeserializer<Location>, JsonSerializer<Location> {

    @Override
    public JsonElement serialize(Location src, Type typeOfSrc, JsonSerializationContext context) {
        return (toJson(src));
    }

    @Override
    public Location deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        return (fromJson(json));
    }

    public static JsonObject toJson(Location location) {
        if (location == null) {
            return (null);
        }

        JsonObject jsonObject = new JsonObject();

        jsonObject.addProperty("world", location.getWorld().getName());
        jsonObject.addProperty("x", location.getX());
        jsonObject.addProperty("y", location.getY());
        jsonObject.addProperty("z", location.getZ());
        jsonObject.addProperty("yaw", location.getYaw());
        jsonObject.addProperty("pitch", location.getPitch());

        return (jsonObject);
    }

    public static Location fromJson(JsonElement jsonElement) {
        if (jsonElement == null || !jsonElement.isJsonObject()) {
            return (null);
        }

        JsonObject jsonObject = jsonElement.getAsJsonObject();

        World world = PotPvPRP.getInstance().getServer().getWorld(jsonObject.get("world").getAsString());
        double x = jsonObject.get("x").getAsDouble();
        double y = jsonObject.get("y").getAsDouble();
        double z = jsonObject.get("z").getAsDouble();
        float yaw = jsonObject.get("yaw").getAsFloat();
        float pitch = jsonObject.get("pitch").getAsFloat();

        return (new Location(world, x, y, z, yaw, pitch));
    }

}

File: src/main/kotlin/net/frozenorb/potpvp/kt/util/serialization/LocationAdapter.kt
package net.frozenorb.potpvp.kt.util.serialization

import net.frozenorb.potpvp.PotPvPRP
import com.google.gson.*
import org.bukkit.Location
import java.lang.reflect.Type

class LocationAdapter : JsonDeserializer<Location>, JsonSerializer<Location> {
    override fun serialize(src: Location, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        return toJson(src) as JsonElement
    }

    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Location? {
        return fromJson(json)
    }

    companion object {
        fun toJson(location: Location?): JsonObject? {
            if (location == null) {
                return null
            }

            val jsonObject = JsonObject()
            jsonObject.addProperty("world", location.getWorld().getName())
            jsonObject.addProperty("x", location.getX() as Number)
            jsonObject.addProperty("y", location.getY() as Number)
            jsonObject.addProperty("z", location.getZ() as Number)
            jsonObject.addProperty("yaw", location.getYaw() as Number)
            jsonObject.addProperty("pitch", location.getPitch() as Number)
            return jsonObject
        }

        fun fromJson(jsonElement: JsonElement?): Location? {
            if (jsonElement == null || !jsonElement.isJsonObject()) {
                return null
            }

            val jsonObject = jsonElement.getAsJsonObject()
            val world = PotPvPRP.getInstance().server.getWorld(jsonObject.get("world").getAsString())
            val x = jsonObject.get("x").getAsDouble()
            val y = jsonObject.get("y").getAsDouble()
            val z = jsonObject.get("z").getAsDouble()
            val yaw = jsonObject.get("yaw").getAsFloat()
            val pitch = jsonObject.get("pitch").getAsFloat()

            return Location(world, x, y, z, yaw, pitch)
        }
    }
}